src/methods.jl

```diff
@@ -10,318 +19,453 @@ derivatives.
+    C₁ = ε * sum(abs, m.coefs)
+    C₂ = M * sum(n -> abs(m.coefs[n] * m.grid[n]^p), eachindex(m.coefs)) / factorial(p)
+    h = convert(T, min((q / (p - q) * C₁ / C₂)^(1 / p), max_step))
 
     # Estimate the accuracy of the method.
-    accuracy = ĥ^(-q) * C₁ + ĥ^(p - q) * C₂
+    accuracy = h^(-q) * C₁ + h^(p - q) * C₂
 
-    # Estimate the value of the derivative.
-    dfdx = sum(i -> convert(T, coefs[i]) * f(T(x + ĥ * grid[i])), eachindex(grid)) / ĥ^q
+    return h, accuracy
+end
 
-    m.history.eps = eps
-    m.history.bound = bound
-    m.history.step = ĥ
-    m.history.accuracy = accuracy
+for direction in [:forward, :central, :backward]
+    fdm_fun = Symbol(direction, "_fdm")
+    grid_fun = Symbol("_", direction, "_grid")
+    @eval begin function $fdm_fun(
+            p::Int,
+            q::Int;
+            adapt::Int=1,
+            condition::Int=DEFAULT_CONDITION,
+            geom::Bool=false
+        )
+            _check_p_q(p, q)
+            grid = collect($grid_fun(p))
+            geom && (grid = _exponentiate_grid(grid))
+            coefs = _coefs(grid, q)
+            return FiniteDifferenceMethod(
+                grid,
+                q,
+                coefs,
+                _make_adaptive_bound_estimator(
+                    $fdm_fun,
+                    p,
+                    adapt,
+                    condition,
+                    geom=geom
+                )
+            )
+        end
 
-    return m, dfdx
+        @doc """
+    $($(Meta.quot(fdm_fun)))(
+        p::Int,
+        q::Int;
+        adapt::Int=1,
+        condition::Int=DEFAULT_CONDITION,
+        geom::Bool=false
+    )
+
+Contruct a finite difference method at a $($(Meta.quot(direction))) grid of `p` linearly
+spaced points.
+
+# Arguments
+- `p::Int`: Number of grid points.
+- `q::Int`: Order of the derivative to estimate.
+
+# Keywords
+- `adapt::Int=1`: Use another finite difference method to estimate the magnitude of the
+    `p`th order derivative, which is important for the step size computation. Recurse
+    this procedure `adapt` times.
+- `condition::Int`: Condition number. See [`DEFAULT_CONDITION`](@ref).
+- `geom::Bool`: Use geometrically spaced points instead of linearly spaced points.
+
+# Returns
+- `FiniteDifferenceMethod`: The specified finite difference method.
+        """ $fdm_fun
+    end
 end
 
-# Handle inputs that aren't `AbstractFloat`s -- assume what you wanted was a `Float64`. This
-# most common way that this method gets invoked is when `x` is an `Integer`.
-function fdm(m::FiniteDifferenceMethod, f, x::Real, val; kwargs...)
-    return fdm(m, f, Float64(x), val; kwargs...)
+function _make_adaptive_bound_estimator(
+    constructor::Function,
+    q::Int,
+    adapt::Int,
+    condition::Int;
+    kw_args...
+)
+    if adapt >= 1
+        estimate_derivative =
+            constructor(q + 1, q, adapt=adapt - 1, condition=condition; kw_args...)
```

Comment:
I think the preferred line breaking is:
```suggestion
```suggestion
        estimate_derivative = constructor(
            q + 1, q, adapt=adapt - 1, condition=condition; kw_args...
        )
```

Rather than breaking after the =

Reply:
Just repeating it here to remove line 330.

```suggestion
        estimate_derivative = constructor(
            q + 1, q, adapt=adapt - 1, condition=condition; kw_args...
        )
```